### PR TITLE
8324995: Shenandoah: Skip to full gc for humongous allocation failures

### PR DESCRIPTION
**Pull request by earthling-amzn**
Shenandoah degenerated cycles do not compact regions. When a humongous allocation fails, it is likely due to fragmentation which is better addressed by a full gc.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blocker
&nbsp;⚠️ Whitespace errors (failed with updated jcheck configuration in pull request)

### Issue
 * [JDK-8324995](https://bugs.openjdk.org/browse/JDK-8324995): Shenandoah: Skip to full gc for humongous allocation failures (**Enhancement** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**)

### Contributors
 * Kelvin Nilsen `<kdnilsen@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17638/head:pull/17638` \
`$ git checkout pull/17638`

Update a local copy of the PR: \
`$ git checkout pull/17638` \
`$ git pull https://git.openjdk.org/jdk.git pull/17638/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17638`

View PR using the GUI difftool: \
`$ git pr show -t 17638`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17638.diff">https://git.openjdk.org/jdk/pull/17638.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17638#issuecomment-1917786873)

### CODE REVIEW

**Comment by earthling-amzn** [comment]
/contributor add kdnilsen

Kelvin developed the original change. I'm just porting it from the shenandoah development repo.

**Review by kdnilsen — APPROVED** [review]

**Review by ysramakrishna — APPROVED** [review]
Changes look great; are there any performance numbers to share for the change?

**Comment by shipilev** [comment]
Well, degenerated cycle can clean up enough space for humongous allocation to succeed? Seems weird to go straight full GC without trying the degen GC first. Is there a substantial performance benefit for doing this?

**Comment by kdnilsen** [comment]
I dusted off the workloads that had originally motivated this change.  Some of the context within which the code executes is different.  At the time, we were doing 64 degens before a full gc.

Here is the test I ran 5 times:
for i in 48g 42g 36g 32g 31g 30g 29g 28g 27g 26g 25g 24g                                                                                                  
do

echo Run TradiShen tip with memory size $i with 4s customer period
>&2 echo Run TradiShen tip with memory size $i with 4s customer period
~/github/jdk.2-1-2024/build/linux-x86_64-server-release/jdk/bin/java \
  -XX:+UnlockExperimentalVMOptions \
  -XX:+UseTransparentHugePages \
  -XX:-ShenandoahPacing \
  -XX:+AlwaysPreTouch -XX:+DisableExplicitGC -Xms$i -Xmx$i \
  -XX:+UseShenandoahGC \
  -Xlog:"gc*=info,ergo" \
  -Xlog:safepoint=trace -Xlog:safepoint=debug -Xlog:safepoint=info \
  -XX:+UnlockDiagnosticVMOptions \
  -jar ~/github/heapothesys/Extremem/target/extremem-1.0-SNAPSHOT.jar \
  -dInitializationDelay=45s -dDictionarySize=16000000 -dNumCustomers=28000000 \
  -dNumProducts=64000 -dCustomerThreads=2000 -dCustomerPeriod=4s -dCustomerThinkTime=1s \
  -dKeywordSearchCount=4 -dServerThreads=5 -dServerPeriod=5s -dProductNameLength=10 \
  -dBrowsingHistoryQueueCount=5 \
  -dSalesTransactionQueueCount=5 \
  -dProductDescriptionLength=64 -dProductReplacementPeriod=25s -dProductReplacementCount=5 \
  -dCustomerReplacementPeriod=30s -dCustomerReplacementCount=1000 -dBrowsingExpiration=1m \
  -dPhasedUpdates=true \
  -dPhasedUpdateInterval=60s \
  -dSimulationDuration=20m -dResponseTimeMeasurements=100000


echo Run Humongous Failure Handling with one degen and memory size $i with 4s customer period
>&2 echo Run Humongous Failure Handling with one degen and memory size $i with 4s customer period
~/gitfarm/shen.humongous-alloc-failure-handling/build/linux-x86_64-server-release/jdk/bin/java \
  -XX:+UnlockExperimentalVMOptions \
  -XX:+UseTransparentHugePages \
  -XX:-ShenandoahPacing \
  -XX:+AlwaysPreTouch -XX:+DisableExplicitGC -Xms$i -Xmx$i \
  -XX:+UseShenandoahGC \
  -Xlog:"gc*=info,ergo" \
  -Xlog:safepoint=trace -Xlog:safepoint=debug -Xlog:safepoint=info \
  -XX:+UnlockDiagnosticVMOptions \
  -jar ~/github/heapothesys/Extremem/target/extremem-1.0-SNAPSHOT.jar \
  -dInitializationDelay=45s -dDictionarySize=16000000 -dNumCustomers=28000000 \
  -dNumProducts=64000 -dCustomerThreads=2000 -dCustomerPeriod=4s -dCustomerThinkTime=1s \
  -dKeywordSearchCount=4 -dServerThreads=5 -dServerPeriod=5s -dProductNameLength=10 \
  -dBrowsingHistoryQueueCount=5 \
  -dSalesTransactionQueueCount=5 \
  -dProductDescriptionLength=64 -dProductReplacementPeriod=25s -dProductReplacementCount=5 \
  -dCustomerReplacementPeriod=30s -dCustomerReplacementCount=1000 -dBrowsingExpiration=1m \
  -dPhasedUpdates=true \
  -dPhasedUpdateInterval=60s \
  -dSimulationDuration=20m -dResponseTimeMeasurements=100000

done

My shen.humongous-alloc-failure-handling branch had an experimental delta from what is here.  The idea was to do only one degen, and then upgrade to full on a humongous alloc failure:

diff --git a/src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp b/src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
index a4b2adc8f5a..42721059b7e 100644
--- a/src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
@@ -108,7 +108,11 @@ void ShenandoahControlThread::run_service() {
 
       // If a humongous allocation has failed, then the heap is likely in need of compaction, so run
       // a full gc (which compacts regions) instead of a degenerated gc (which does not compact regions).
-      if (ShenandoahDegeneratedGC && heuristics->should_degenerate_cycle() && !humongous_alloc_failure_pending) {
+
+      // Experiment: If we had a humongous_alloc_failure, make sure we try at least one degen before going to full.
+      if (ShenandoahDegeneratedGC &&
+          ((humongous_alloc_failure_pending && heap->shenandoah_policy()->consecutive_degenerated_gc_count() == 0) ||
+           (!humongous_alloc_failure_pending && heuristics->should_degenerate_cycle()))) {
         heuristics->record_allocation_failure_gc();
         policy->record_alloc_failure_to_degenerated(degen_point);
         mode = stw_degenerated;

The results are summarized in the attached spreadsheet.
[humongous-alloc-failure-handling.xlsx](https://github.com/openjdk/jdk/files/14173614/humongous-alloc-failure-handling.xlsx)

**Comment by kdnilsen** [comment]
Analysis of the spreadsheet results:

1. I repeated the experiment 5 times because the results were so "noisy"
2. Turns out, the results have almost nothing to do with how many degens we do before upgrading to full following a humongous allocation failure.  The reaons is because humongous allocation failures are very rare (on these workloads).
3. The biggest problem is "upgrading degen to full gc".  Whenever we see more than a few upgrades to full gc, the P50 latency explodes on us.
   1. Often, we see the same exat workload configuration run without any upgrades to full gc and perform much better.  Consider, for example 25g heap, Mainline.  Three runs ran without upgrades to full gc.  P50 latency was 1_147, 1_145, and 1_144 microseconds respectively.  Two runs experienced upgrades to full gc.  Those had p50 latencies of 223_236_560 and 86_091_599 microseconds respectively.
   2. Note that there's not much middle ground here.  If the workload avoids upgrading to full, the results are good.  If it experiences a single upgrade to full, it is likely to spiral into a lot more upgrades to full.
   3. My assessment: upgrade to full is generally counterproductive.  We take a "long pause" to do degenerated work, and then we throw that work away and start all over with another even longer pause to do full gc.   Meanwhile, client requests continue to accumulate.
   
   
Based on these measurements, I'm inclined to recommend the following:

1. If a humongous allocation request fails, we should ask ourselves "How much humongous memory was available at the most recent freeset rebuild?"  If that amount of memory is greater or equal to the size of the requested humongous allocation, we should degenerate (always).  If it is smaller, we should do Full GC.

2. We should never upgrade to full from degen.  The heuristic about "non-productive" degen is confusing us.  What typically happens is we get into an unproductive cycle consisting of the following:

   1. Concurrent GC fails to allocate
   2. Degenerated GC takes multiple seconds to complete
   3. We resume execution with multiple seconds of pentup demand for allocation
   4. Concurrent GC triggers immediately, but may degenerate again if we are "near the edge of the cliff" due to the burst of "catch-up" work that we are trying to perform
   5. All the memory just allocated during concurrent GC is "floating garbage".  None of it can be reclaimed by degenerated GC.
   6. Degen is unproductive.
   7. We upgrade to Full GC.
   8. Full GC reclaims all the floating garbage.
   9. When we resume mutator work, the pentup demand is even larger than the previous scenario because we now have pent up demand from both STW degen phase and STW full GC phase.
   10. So we repeat this cycle over and over and over again.
   
 I'm going to try an experiment where the only time we do full gc is if there is humongous allocation failure and humongous memory from most recently built free set is too small to satisfy the alloc request.

**Comment by kdnilsen** [comment]
Out of cycle alloc failure might as well do full GC.  I would keep that behavior in my experiment.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 1]
```diff
@@ -481,8 +485,9 @@ void ShenandoahControlThread::handle_alloc_failure(ShenandoahAllocRequest& req,
 
 void ShenandoahControlThread::handle_alloc_failure_evac(size_t words) {
   ShenandoahHeap* heap = ShenandoahHeap::heap();
+  bool is_humongous = words > ShenandoahHeapRegion::humongous_threshold_words();
 
-  if (try_set_alloc_failure_gc()) {
+  if (try_set_alloc_failure_gc(is_humongous)) {
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:490]: I think we can replace this with assert(words <= ShenandoahHeapRegion::humongous_ghreshold_words()

We do not evacuate humongous objects.

**Comment by kdnilsen** [comment]
There is a separate issue going that has been uncovered by testing of this PR.  We have observed that VM_HandshakeAllThreads::doit() very rarely iterates very slowly over threads.  This has been observed with HandShakeForDeflation causing a pause of over 280s.  I will file a JBS ticket for this.

**Comment by kdnilsen** [comment]
Meanwhile, I've done some analysis of the attached GC log to better understand tradeoffs between doing degenerated GC vs. Full GC when mutator experiences out-of-memory during allocation request.  The command used for this run is:

echo Run TradiShen tip with memory size $i with 4s customer period
>&2 echo Run TradiShen tip with memory size $i with 4s customer period
~/github/jdk.2-1-2024/build/linux-x86_64-server-release/jdk/bin/java \
  -XX:+UnlockExperimentalVMOptions \
  -XX:+UseTransparentHugePages \
  -XX:-ShenandoahPacing \
  -XX:+AlwaysPreTouch -XX:+DisableExplicitGC -Xms$i -Xmx$i \
  -XX:+UseShenandoahGC \
  -Xlog:"gc*=info,ergo" \
  -Xlog:safepoint=trace -Xlog:safepoint=debug -Xlog:safepoint=info \
  -XX:+UnlockDiagnosticVMOptions \
  -jar ~/github/heapothesys/Extremem/target/extremem-1.0-SNAPSHOT.jar \
  -dInitializationDelay=45s -dDictionarySize=16000000 -dNumCustomers=28000000 \
  -dNumProducts=64000 -dCustomerThreads=2000 -dCustomerPeriod=4s -dCustomerThinkTime=1s \
  -dKeywordSearchCount=4 -dServerThreads=5 -dServerPeriod=5s -dProductNameLength=10 \
  -dBrowsingHistoryQueueCount=5 \
  -dSalesTransactionQueueCount=5 \
  -dProductDescriptionLength=64 -dProductReplacementPeriod=25s -dProductReplacementCount=5 \
  -dCustomerReplacementPeriod=30s -dCustomerReplacementCount=1000 -dBrowsingExpiration=1m \
  -dPhasedUpdates=true \
  -dPhasedUpdateInterval=60s \
  -dSimulationDuration=20m -dResponseTimeMeasurements=100000

with $i equal to 27g
[capture.27g.tradishen.log](https://github.com/openjdk/jdk/files/14213319/capture.27g.tradishen.log)

**Comment by kdnilsen** [comment]
Some key observations, which are highlighted in the attached excel spreadsheet:

1. Concurrent GCs take roughly twice as long as degenerated GCs, and use half as many worker threads. (out-of-cycle degen: 3.105s, fully conc gcs: 6.44s and 6.341s)
2. The sum of degen time plus 1/2 conc gc time that precedes degen is nearly constant: 3.368s, with std dev: 0.291
3. Note that a longer concurrent GC effort is accompanied by a shorter degen effort because degen GC leverages progress already made by concurrent GC.
4. Full GC requires 50% more time than a degen GC (5.026s avg vs 3.368s avg).  This is presumably because full GC compacts everything, whereas degen GC uses garbage-first heuristics to only evacuate regions that are "convenient".
5. Full GC typically reclaims 88% more garbage than conc gc and degen.  This is presumably because full GC reclaims all the floating garbage for allocations that occur following the start of a concurrent GC effort.
6. When Full GC is preceded by a failed concurrent effort and then a failed degen effort, the total duration of the GC cycle is much longer than a typical degen cycle.  The time for a conc/degen cycle ranges from 3.105s (for out-of-cycle degen with no concurrent phase) to 6.44s for concurrent phase without any degen.  In contrast, duration of a full GC cycle ranges from 7.431 w (when 2.393s of concurrent gc effort transitions directly to Full) to 10.329s (when 2.483s of conc gc is followed by 2.636s of degen gc, which experiences bad progress and upgrades to full gc).
7. Under heavy allocation load, full GC tends to self-perpetuate.  This is caused by multiple factors:
    1. The very long GC duration means we're reclaiming memory less efficiently, even though each full GC yields greater available memory
    2. Full GCs, on average, are yielding 1,043 MB/s of wall-clock time spent in GC compared to 931 MB/s for concurrent and degen GC.  However, degen offers a higher peak yield of 1,770 MB/s compared to the peak yield of 1,181 MB/s for full GC.
    3. In terms of CPU time dedicated to GC, Full GC averages 1,239 MB/s vs. 1,529 MB/s for degen.  Peak performance also favors degen, with max of 2,029 MB/s vs. full GC max of 1,460 MB/s.
    4. A final strike against Full GC is that it leaves the heap in a state that is very difficult to recover from.  Specifically, following Full GC, there is no garbage in the heap.  If we immediately trigger concurrent GC, it will be unproductive because there is no garbage to be found in the heap, and any floating garbage created following the start of concurrent GC will not be found until the next concurrent GC cycle.  This is observed in the log, with GC(85) through GC(91) each upgrading to Ful
[full-vs-degen.xlsx](https://github.com/openjdk/jdk/files/14213994/full-vs-degen.xlsx)
l GC.

**Comment by earthling-amzn** [comment]
So, it sounds like there is not much value in going straight to a full GC? I don't think we can really know ahead of time whether or not a degenerated cycle will free up enough contiguous regions to satisfy the humongous allocation request. We could make an educated guess based on heap fragmentation and occupancy.

**Comment by kdnilsen** [comment]
I ran an experiment in which I discourage full GC.  Attached is the diff from this branch, and an spreadsheet summarizing the performance results.  HIghlights:

1. I overrode the default ShenandoahFullGCThreshold with 64 on the command line.
2. should_degenerate_cycle() only returns false if a humongous alloc failed and the cumulative requested humongous size > the amount of humongous memory available at the most recent free-set rebuild operation, or if consecutive_degen_count > ShenandoahFullGCThreshold.  (This mostly preserves existing behavior, but gives users the option of increasing ShenandoahFullGCThreshold as I did.)
3. Across almost all latency metrics, the new code performs better than original code.
4. Full GCs decreased to a maximum of 4 on any run, whereas the original max full gc was 104.  Average number of full GCs decreased to ~3 from typically more than 50.
5. We do quite a few more degens and fewer successful concurrent GCs.
6. More testing should be done with more diverse workloads, and after we fix the existing problem with VM_HandshakeAllThreads::do_it().

See attached diff file and excel spreadhsheet for more detail.

[discourage-full-experiment.xlsx](https://github.com/openjdk/jdk/files/14215043/discourage-full-experiment.xlsx)

**Comment by kdnilsen** [comment]
[discourage-full-experiment.diff.txt](https://github.com/openjdk/jdk/files/14215058/discourage-full-experiment.diff.txt)

**Comment by earthling-amzn** [comment]
Not sure I understand:
> cumulative requested humongous size > the amount of humongous memory available at the most recent free-set rebuild operation

Don't we already know that there isn't enough contiguous memory for _this_ request? Wouldn't the value of the accumulated failed humongous request sizes be greater than the size for the _current_ request?

**Comment by kdnilsen** [comment]
There isn't enough memory right now.  But there may have been at the end of the most recent GC.  The question: Is a normal GC likely to reclaim enough contiguous memory to satisfy the humongous allocation request.  If the previous normal GC was successful, then the new one is also likely to be successful.

**Comment by ysramakrishna** [comment]
In regards to:

> ... that the space will reappear in contiguous form as soon as the ongoing concurrent gc (albeit degenerated now?) completes without taking recourse to a full gc.

I may be misunderstanding here, but what if the failure to honor the request doesn't cause degeneration but is held up while the concurrent cycle completes, after which said space may become available, albeit penalizing this allocation request with a much longer stall, while not subjecting other threads to the pause. May be that isn't worthwhile, or would require a bigger change.

**Comment by kdnilsen** [comment]
Thanks Ramki.  I agree with your "rewording" of "what I am saying".

If anything, the changes proposed in https://github.com/openjdk/jdk/pull/17561 will reduce the likelihood that we experience a humongous alloc failure, and in the rare case that we do, will increase the likelihood that normal concurrent GC (with degen if necessary) will recover sufficient contiguous memory to satisfy the humongous allocation request.

I do agree that there's further room for smart adaptive behavior as you suggest.

**Comment by kdnilsen** [comment]
Ramki: what you are suggesting is what has been described as ShenandoahPacing and/or Throttling.  That can avoid the need for degeneration, especially when we surge the number of worker threads, for example.

**Comment by ysramakrishna** [comment]
> If anything, the changes proposed in #17561 will reduce the likelihood that we experience a humongous alloc failure, and in the rare case that we do, will increase the likelihood that normal concurrent GC (with degen if necessary) will recover sufficient contiguous memory to satisfy the humongous allocation request.

Does it seem worthwhile then to run perf numbers with these changes along with the changes in https://github.com/openjdk/jdk/pull/17561 so as to get sharper performance numbers that include both changes (as well as each individually as you did here and in the other one). Each seems worthwhile on its own, and I am curious of the composition of the two...

**Comment by kdnilsen** [comment]
Agree with testing combination of capabilities.  Also want to get the handshake issue out of the picture.  I'm looking at that.

**Comment by earthling-amzn** [comment]
Abandoning this PR while we conduct further testing.

**Pull request closed without merging**